auxx/src/Repl.hs

```diff
@@ -0,0 +1,178 @@
+        -- The node (and the auxx plugin) should use this action exclusively for printing.
+        --
+        -- This action blocks until Haskeline is initialized (happens almost immediately).
+        --
+    }
+
+-- | The action to post a command to be executed by the auxx plugin and wait for
+-- its result. This action blocks until the result is available: that is, until
+-- command execution completes or an exception is thrown.
+type PutCommandAction = Command -> AuxxRepl CommandResult
+
+-- | Create the auxx REPL (as an abstract @'IO' ()@ action) and the interface to
+-- interact with it ('WithCommandAction'). You are supposed to fork the REPL
+-- action into a separate thread: 'withAuxxRepl' does it for you, so use it
+-- instead unless you need to add special logic.
+createAuxxRepl :: (MonadIO m, Mockable Catch m) => IO (WithCommandAction m, IO ())
+createAuxxRepl = do
+    nextCommandVar <- newEmptyMVar
+    lastResultVar <- newEmptyMVar
+    printActionVar <- newEmptyMVar
+    let
+        withCommand cont = do
+            cmd <- takeMVar nextCommandVar
+            res <- handle
+                (\e -> return $ CommandFailure e)
+                (CommandSuccess <$ cont cmd)
+            putMVar lastResultVar res
+        getPrintAction = readMVar printActionVar
+        putCommand cmd = do
+            putMVar nextCommandVar cmd
+            takeMVar lastResultVar
+    return (WithCommandAction{..}, auxxReplOn printActionVar putCommand)
+
+-- | Fork a REPL and provide 'WithCommandAction' to a continuation. Under the
+-- hood, 'withAuxxRepl' uses 'race_' from the @async@ package, which means that
+-- when either the REPL or the continuation completes (or throws an exception),
+-- 'withAuxxRepl' as a whole completes. Assuming we run the node in the
+-- continuation, using 'race_' has the desired consequence that in order to quit
+-- auxx, we merely need to end the REPL, and the node will be killed
+-- automatically (and vice versa, if the node dies for some reason, the user
+-- won't be left in an unusable REPL).
+withAuxxRepl :: (MonadIO m, Mockable Catch m) => (WithCommandAction m -> IO ()) -> IO ()
+withAuxxRepl action = do
+    (c, auxxRepl) <- createAuxxRepl
+    race_ auxxRepl (action c)
+
+-- Internal function, not exported.
+-- | Initialize Haskeline, put the 'PrintAction' to the provided MVar (assumes
+-- the MVar is empty), and run the REPL.
+auxxReplOn :: MonadIO m => MVar (PrintAction m) -> PutCommandAction -> IO ()
+auxxReplOn printActionVar putCommand = Haskeline.runInputT settings $ do
+    printAction <- Haskeline.getExternalPrint
+    putMVar printActionVar (liftIO . printAction . toString)
+    greetings
+    loop (prompt putCommand)
+  where
+    greetings = do
+        outputTextLn "Welcome to Cardano SL Auxx REPL"
+        outputTextLn helpMessage
+    settings = Haskeline.Settings
+        { Haskeline.complete       = Haskeline.noCompletion
+        , Haskeline.historyFile    = Nothing -- TODO: pass it from CLI
```

Comment:
I created https://issues.serokell.io/issue/CSL-1779
We can do it later, but can you add issue id to this TODO?

Reply:
done